File: src/java/foundgine-core/src/main/java/com/foundgine/core/semantic/planning/RelationshipTraversalOptimizationRule.java
package com.foundgine.core.semantic.planning;

import com.foundgine.core.semantic.RelationshipCardinality;
import java.util.*;

/** Adds a cardinality-aware traversal hint to relationship nodes. */
public final class RelationshipTraversalOptimizationRule implements IPlanRewriteRule {
    @Override public String name() { return "relationship.traversal.strategy"; }
    @Override public List<String> preconditions() { return List.of("plan contains relationship traversal nodes", "relationship cardinality metadata is available"); }
    @Override public List<String> securityObligations() { return List.of("authorization.required", "visibility.relationship", "planning.cache-context-isolation"); }
    @Override public double costImpact() { return .5d; }
    @Override public double benefitEstimate() { return 1.5d; }
    @Override public int priority() { return 20; }
    @Override public boolean canApply(SemanticPlan plan) { Objects.requireNonNull(plan); return containsEligible(plan.root()); }
    @Override public SemanticPlan apply(SemanticPlan plan) {
        Objects.requireNonNull(plan); if (!canApply(plan)) return plan;
        boolean[] changed = {false}; var root = rewrite(plan.root(), changed);
        return changed[0] ? new SemanticPlan(root, plan.requiredSecurityInvariants(), plan.authorizationBinding()) : plan;
    }
    private static SemanticPlanNode rewrite(SemanticPlanNode node, boolean[] changed) {
        var mode = node.traversalMode();
        if (node.viaRelationship() != null && node.relationshipCardinality() != null) {
            var target = node.relationshipCardinality() == RelationshipCardinality.ONE ? RelationshipTraversalMode.SINGLE_HOP : RelationshipTraversalMode.SET_BASED;
            if (mode != target) { mode = target; changed[0] = true; }
        }
        List<SemanticPlanNode> children = new ArrayList<>(node.children().size()); boolean childChanged = false;
        for (var child : node.children()) { var r = rewrite(child, changed); children.add(r); childChanged |= r != child; }
        if (childChanged) changed[0] = true;
        if (mode != node.traversalMode() || childChanged)
            return new SemanticPlanNode(node.id(), node.operation(), node.entityId(), node.fields(), node.viaRelationship(), node.viaConnection(), children, node.queryOptions(), node.authorization(), node.relationshipCardinality(), mode, node.traversalOrder(), node.aggregateExecutionStrategy());
        return node;
    }
    private static boolean containsEligible(SemanticPlanNode node) {
        return (node.viaRelationship() != null && node.relationshipCardinality() != null) || node.children().stream().anyMatch(RelationshipTraversalOptimizationRule::containsEligible);
    }
}

File: src/csharp/Foundgine.Core/Semantic/Planning/RelationshipTraversalOptimizationRule.cs
using Foundgine.Core.Semantic;

namespace Foundgine.Core.Semantic.Planning;

/// <summary>
/// Adds a cardinality-aware traversal hint to relationship nodes.
/// One-to-one traversals are marked as SingleHop; collection traversals are
/// marked SetBased. The rule never changes topology, fields, filters,
/// authorization, pagination, or relationship identity. Providers may use the
/// hint to choose a cheaper physical traversal strategy when they can prove it
/// is safe.
/// </summary>
public sealed class RelationshipTraversalOptimizationRule : IPlanRewriteRule
{
    public string Name => "relationship.traversal.strategy";

    public IReadOnlyList<string> Preconditions =>
    [
        "plan contains relationship traversal nodes",
        "relationship cardinality metadata is available"
    ];

    public IReadOnlyList<string> SecurityObligations =>
    [
        "authorization.required",
        "visibility.relationship",
        "planning.cache-context-isolation"
    ];

    public double CostImpact => 0.5d;
    public double BenefitEstimate => 1.5d;
    public bool IsIdempotent => true;
    public int Priority => 20;

    public bool CanApply(SemanticPlan plan)
    {
        ArgumentNullException.ThrowIfNull(plan);
        return ContainsEligible(plan.Root);
    }

    public SemanticPlan Apply(SemanticPlan plan)
    {
        ArgumentNullException.ThrowIfNull(plan);
        if (!CanApply(plan))
            return plan;

        var changed = false;
        var root = Rewrite(plan.Root, ref changed);
        return changed ? new SemanticPlan(root, plan.RequiredSecurityInvariants, plan.AuthorizationBinding) : plan;
    }

    private static SemanticPlanNode Rewrite(SemanticPlanNode node, ref bool changed)
    {
        var mode = node.TraversalMode;
        if (node.ViaRelationship is not null && node.RelationshipCardinality is { } cardinality)
        {
            var target = cardinality == RelationshipCardinality.One
                ? RelationshipTraversalMode.SingleHop
                : RelationshipTraversalMode.SetBased;
            if (mode != target)
            {
                mode = target;
                changed = true;
            }
        }

        var children = new SemanticPlanNode[node.Children.Count];
        var childrenChanged = false;
        for (var i = 0; i < node.Children.Count; i++)
        {
            children[i] = Rewrite(node.Children[i], ref changed);
            childrenChanged |= !ReferenceEquals(children[i], node.Children[i]);
        }

        if (childrenChanged)
            changed = true;

        return changed
            ? node with { TraversalMode = mode, Children = children }
            : node;
    }

    private static bool ContainsEligible(SemanticPlanNode node) =>
        (node.ViaRelationship is not null && node.RelationshipCardinality is not null) ||
        node.Children.Any(ContainsEligible);
}
